app/src/services/tag.js

```diff
@@ -7,42 +7,196 @@ const { Tag, VersionTag } = require('../db/models');
+   */
+  dissociateTags: async (versionId, keys = undefined, etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Tag.startTransaction();
+      let response = [];
+
+      // delete joining records
+      response = await VersionTag.query(trx)
+        .allowGraph('tag')
+        .withGraphJoined('tag')
+        .whereIn('key', keys)
+        .modify('filterVersionId', versionId)
+        .delete();
+
+      // delete all orphaned tags
+      await service.pruneOrphanedTags(trx);
+
+      if (!etrx) await trx.commit();
+      return Promise.resolve(response);
+    } catch (err) {
+      if (!etrx && trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  /**
+   * @function deleteOrphanedTags
+   * deletes Tag records if they are no longer related to any versions
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<number>} The result of running the delete operation (number of rows deleted)
+   * @throws The error encountered upon db transaction failure
+   */
+  pruneOrphanedTags: async (etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Tag.startTransaction();
+
+      const deletedTagIds = await Tag.query(trx)
+        .allowGraph('versionTag')
+        .withGraphJoined('versionTag')
+        .select('tag.id')
+        .whereNull('versionTag.tagId');
+
+      const response = await Tag.query(trx)
+        .delete()
+        .whereIn('id', deletedTagIds.map(({ id }) => id));
+
+      if (!etrx) await trx.commit();
+      return Promise.resolve(response);
+    } catch (err) {
+      if (!etrx && trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  /**
+   * @function createTags
+   * Inserts any tag records if they dont already exist in db
+   * @param {object} tags Incoming object with `<key>:<value>` tags to add for this version
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<object>} an array of all input tags
+   * @throws The error encountered upon db transaction failure
+   */
+  createTags: async (tags, etrx = undefined) => {
+    let trx;
+    let response = [];
+    try {
+      trx = etrx ? etrx : await Tag.startTransaction();
+
+      // get all tags already in db
+      const allTags = await Tag.query(trx).select();
+      const existingTags = [];
+      const newTags = [];
+
+      tags.forEach(({ key, value }) => {
+        // if tag is already in db
+        if (allTags.some(tag => (tag.key === key.toLowerCase() && tag.value === value))){
+          existingTags.push({ id: allTags.find(tag => tag.key === key).id, key: key, value: value });
+        }
+        // else add to array for inserting
+        else {
+          newTags.push({ key: key.toLowerCase(), value: value });
+        }
+      });
+
+      // insert new tags
+      if (newTags.length) {
+        const newTagArr = await Tag.query(trx).insert(newTags).returning('*');
```

Comment:
Multiline this query. Also worthwhile naming this variable as newTagset or something - always want to describe what the array is about instead of calling the variable an array itself.